src/main/java/graphql/schema/transform/FieldVisibilitySchemaTransformation.java

```diff
@@ -0,0 +1,206 @@
+package graphql.schema.transform;
+
+import graphql.PublicApi;
+import graphql.schema.GraphQLEnumType;
+import graphql.schema.GraphQLFieldDefinition;
+import graphql.schema.GraphQLInputObjectField;
+import graphql.schema.GraphQLInterfaceType;
+import graphql.schema.GraphQLNamedSchemaElement;
+import graphql.schema.GraphQLObjectType;
+import graphql.schema.GraphQLSchema;
+import graphql.schema.GraphQLSchemaElement;
+import graphql.schema.GraphQLType;
+import graphql.schema.GraphQLTypeVisitorStub;
+import graphql.schema.GraphQLUnionType;
+import graphql.schema.SchemaTraverser;
+import graphql.schema.transform.VisibleFieldPredicateEnvironment.VisibleFieldPredicateEnvironmentImpl;
+import graphql.util.TraversalControl;
+import graphql.util.TraverserContext;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static graphql.schema.SchemaTransformer.transformSchema;
+import static graphql.util.TreeTransformerUtil.deleteNode;
+
+/**
+ * Transforms a schema by applying a visibility predicate to every field.
+ */
+@PublicApi
+public class FieldVisibilitySchemaTransformation {
+
+    private final VisibleFieldPredicate visibleFieldPredicate;
+
+    public FieldVisibilitySchemaTransformation(VisibleFieldPredicate visibleFieldPredicate) {
+        this.visibleFieldPredicate = visibleFieldPredicate;
+    }
+
+    /**
+     * Before and after callbacks useful for side effects (logs, stopwatches etc).
+     */
+    protected void beforeTransformation() {
```

Comment:
can we make this hooks (just Function/Consumer) is enough. 

I think that is cleaner than asking people to make subclasses

Reply:
Updated.